src/ipc/dbus/DBusObjectRegistry.cpp

```diff
@@ -78,44 +106,110 @@ void DBusObjectRegistry::unregisterObject(const QString &objectPath)
 {
     init();
     auto serviceName = DBusManager::instance().serviceName();
-    if (m_master) {
+    if (isMaster()) {
         m_master->unregisterObject(objectPath, serviceName);
     } else {
         return m_objectRegistryAsyncProxy->unregisterObject(objectPath, serviceName);
     }
 }
 
-const QMap<QString, QString> &DBusObjectRegistry::objects(bool blocking)
+void DBusObjectRegistry::updateObjects(const QMap<QString, QString>& objectMap)
+{
+    int version = objectMap[versionKey].toInt(); // 0 if key is not present
+    if (version && (version > m_registryVersion ||
+            ((std::numeric_limits<int>::max() - m_registryVersion < 100) && (version < 100)))) // version number overflow
+    {
+        m_registryVersion = version;
+        m_objects = objectMap;
+        m_objects.remove(versionKey);
+        emit objectsChanged();
+    }
+}
+
+void DBusObjectRegistry::refreshObjects()
+{
+    ObjectRegistryIPCDBusProxy objectRegistryProxy;
+    objectRegistryProxy.ipc()->setServiceName(m_serviceName);
+    objectRegistryProxy.connectToServer();
+    updateObjects(objectRegistryProxy.getObjects());
+}
+
+const QMap<QString, QString>& DBusObjectRegistry::objects(bool blocking)
 {
     init();
-    if (m_master) {
-        return m_master->objects();
+    if (!isMaster() && blocking && m_registryVersion == invalidRegistryVersion) {
+        refreshObjects();
+    }
+    return m_objects;
+}
+
+void DBusObjectRegistry::onObjectAdded(const QString &objectPath, const QString &serviceName, int registryVersion)
+{
+    if (hasValidObjects() && registryVersion == nextVersion(m_registryVersion)) {
+        m_registryVersion = registryVersion;
+        if (!m_objects.contains(objectPath)) {
+            m_objects.insert(objectPath, serviceName);
+            emit objectsChanged();
+        } else {
+            qCCritical(LogIpc) << "Cannot add object with object path:" << objectPath << ", service name:" << serviceName
+                               << "Object registry already contains this object path with service name:" << m_objects[objectPath];
+        }
+    } else if (registryVersion != version()) {
+        refreshObjects(); // critical situation, intermediate version is missing. Force update
```

Comment:
Do you really intend to initiate a blocking IPC call here ?

Reply:
1) this is unlikely to happen. If that happens, we missed DBus update signal somehow, that means critical error. This could potentially go as async request, but async may be actually less reliable in this place. 
2) it already comes from dbus update signal, so it should not block anything important